File: sol/solution/0001-0100/0023. Merge k Sorted Lists/Solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        val pq = java.util.PriorityQueue<ListNode>(compareBy { it.`val` })
        for (node in lists) {
            if (node != null) pq.add(node)
        }
        val dummy = ListNode(0)
        var curr = dummy
        while (pq.isNotEmpty()) {
            val node = pq.poll()
            curr.next = node
            curr = curr.next!!
            if (node.next != null) pq.add(node.next)
        }
        return dummy.next
    }
}

File: sol/solution/0001-0100/0023. Merge k Sorted Lists/Solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    if (!lists || lists.length === 0) {
        return null;
    }
    return mergeKListsHelper(lists, 0, lists.length - 1);
};
var mergeKListsHelper = function(lists, left, right) {
    if (left === right) {
        return lists[left];
    }
    var mid = Math.floor((left + right) / 2);
    var l1 = mergeKListsHelper(lists, left, mid);
    var l2 = mergeKListsHelper(lists, mid + 1, right);
    return mergeTwoLists(l1, l2);
};
var mergeTwoLists = function(l1, l2) {
    var dummy = new ListNode();
    var current = dummy;
    while (l1 !== null && l2 !== null) {
        if (l1.val < l2.val) {
            current.next = l1;
            l1 = l1.next;
        } else {
            current.next = l2;
            l2 = l2.next;
        }
        current = current.next;
    }
    current.next = l1 !== null ? l1 : l2;
    return dummy.next;
};